ginkgobioworks/react-json-schema-form-builder
# Migrate from Flow to Typescript #456

Closes #456 

This has been a fun challenge and a great TypeScript exercise! 😄

It would be great to get another pair of eyes on this work to ensure nothing was missed. This touches the entire library, so it is a pretty major change. I am happy to update any issues you find. I have also checked "Allow edits by maintainers" so feel free to add to my branch if necessary.

⚠️  **Of note:** I have not been able to test importing the TS version of this library via npm. It may be worth releasing a beta version to test thoroughly before the official release.

**What was done:**
- Converted all Flow syntax to TS
- Removed flow dependencies and files
- Installed any necessary libraries for type definitions
- Fixed any syntax that was not compatible with TypeScript
- Converted all unit tests to TS and updated Jest configuration
- Noted possible issues with `TODO` comments (will also create separate issues for each)
- Added package scripts to run TS checks in testing and on deployment
- Updated documentation, including CONTRIBUTING guide to reflect changes


**To test**

1. Pull this working branch
2. Run `npm i` to install dependencies in both the root folder and `example/`
3. You should be able to run the example site locally as usual with `npm start`
4. You can run typescript checks with the new script: `npm run test:ts` on the project root, and `npm run typescript` on the example site. Note: this will not emit any files, but will only check for errors.
5. All unit tests should work as they do on the main branch (note: there are some warnings about not wrapping tests in `act()` but this is a pre-existing issue and not a result of the TypeScript changes.
6. Building the project will compile into `index.js` and `index.modern.mjs` as usual, however you will also notice extra TypeScript definition files, and those will be exported in a directory structure which matches the `src` directory.
